### PR TITLE
NopResource added.

### PR DESCRIPTION
**Pull request by rahulgrover99**
Signed-off-by: Rahul Grover <rahulgrover99@gmail.com>

### CODE REVIEW

File: internal/controller/nopresource/nopresource.go [hunk 1]
```diff
@@ -128,13 +136,22 @@ type external struct {
...
+	fmt.Printf(cr.Spec.ForProvider.ConditionAfter[0].Condition)
+	fmt.Printf("\n\n")
+
+	if meta.WasDeleted(mg) {
+		return managed.ExternalObservation{ResourceExists: false}, nil
 	}
 
 	// These fmt statements should be removed in the real implementation.
 	fmt.Printf("Observing: %+v", cr)
+	cr.SetConditions(runtimev1alpha1.Available())
```
Comment (hasheddan) [inline internal/controller/nopresource/nopresource.go]: @rahulgrover99 are you not seeing this reflected on managed resource despite reaching this point in the controller?

Comment (rahulgrover99) [reply internal/controller/nopresource/nopresource.go]: Yes, I am not able to see the `Status` and `State` when I run `k get nopresource`. 
However I observed when I run `k describe nopresource`, I see those fields populated. 
```
  At Provider:
  Conditions:
    Last Transition Time:  2021-04-22T14:19:35Z
    Reason:                Available
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-04-22T14:19:35Z
    Reason:                ReconcileSuccess
    Status:                True
    Type:                  Synced
Events:                    <none>
```

File: apis/sample/v1alpha1/types.go [hunk 2]
```diff
@@ -22,50 +22,57 @@ import (
...
+	AtProvider          NopResourceObservation `json:"atProvider,omitempty"`
 }
 
 // +kubebuilder:object:root=true
 
-// A MyType is an example API type.
+// A NopResource is an example API type.
 // +kubebuilder:subresource:status
 // +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.bindingPhase"
 // +kubebuilder:printcolumn:name="STATE",type="string",JSONPath=".status.atProvider.state"
```
Comment (hasheddan) [inline apis/sample/v1alpha1/types.go]: @rahulgrover99 these need to be updated to point to correct fields on the status of this resource and then they will show up with `k get` :)

File: internal/controller/nopresource/nopresource.go [hunk 3]
```diff
@@ -128,13 +97,70 @@ type external struct {
...
+		return managed.ExternalObservation{}, errors.New(errNotNopResource)
+	}
+	startTime := cr.CreationTimestamp
+
+	// If object was deleted, return it does not exist so that managed reconciler removes finalizer
+	if meta.WasDeleted(mg) {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
+
+	// Logic 1 - Set condition as soon as the time elapsed in seconds is equal to time specified in Spec
```
Comment (rahulgrover99) [inline internal/controller/nopresource/nopresource.go]: @hasheddan this logic will set the condition as soon as the time elapsed in seconds is equal to time specified in Spec
The behavior of the same can be found at `logs/logic1. log`

File: internal/controller/nopresource/nopresource.go [hunk 4]
```diff
@@ -128,13 +97,70 @@ type external struct {
...
+	// 			LastTransitionTime: metav1.Now(),
+	// 			Reason:             runtimev1alpha1.ReasonAvailable,
+	// 		}
+
+	// 		cr.Status.SetConditions(x)
+
+	// 	}
+	// }
+
+	// Logic 2 - Set condition at every reconcile to the latest specified condition in Spec
```
Comment (rahulgrover99) [inline internal/controller/nopresource/nopresource.go]: This logic will set conditions at every reconcile to the latest specified condition in Spec. 
The behavior of the same can be found at logs/logic2. log

File: logs/logic1. log [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+
+2021-04-27T11:18:00.593Z        DEBUG   provider-template       External resource is up to date {"controller": "managed/nopresource.sample.template.crossplane.io", "request": "/example", "uid": "94b97e85-99ff-4b3c-8f2d-fc81ae3e3716", "version": "1498873", "external-name": "example", "requeue-after": "2021-04-27T11:18:01.593Z"}
+2021-04-27T11:18:01.599Z        DEBUG   provider-template       Reconciling     {"controller": "managed/nopresource.sample.template.crossplane.io", "request": "/example"}
+
+
+        My values
+
+17.599209698s
+
+ReconcileSuccess  True Synced
```
Comment (rahulgrover99) [inline logs/logic1. log:310]: The major difference in the 2 behaviors can be seen after 15s.

File: logs/logic2.log [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+2021-04-27T11:56:21.455Z        DEBUG   provider-template       External resource is up to date     {"controller": "managed/nopresource.sample.template.crossplane.io", "request": "/example", "uid": "9ed9ba34-7b68-429c-abf5-4ba159f8e917", "version": "1507334", "external-name": "example", "requeue-after": "2021-04-27T11:56:22.455Z"}
+2021-04-27T11:56:21.460Z        DEBUG   provider-template       Reconciling     {"controller": "managed/nopresource.sample.template.crossplane.io", "request": "/example"}
+Calling update on index 2
+
+
+        My values
+
+16.460659897s
+
+Available  False Synced
```
Comment (rahulgrover99) [inline logs/logic2.log:285]: Contrary to the previous behavior it will hold these conditions.

**Review by rahulgrover99 — COMMENTED** [review]
@hasheddan Which one of these should be the expected behavior from the provider?

File: logs/logic1. log (hunk 5 above)
Comment (rahulgrover99) [reply logs/logic1. log:310]: The output you are seeing above is `fmt.Printf("%s %s %s %s", string(e.Reason), string(e.Message), string(e.Status), string(e.Type))`

File: internal/controller/nopresource/nopresource.go [hunk 7]
```diff
@@ -217,7 +189,29 @@ func (c *external) Delete(ctx context.Context, mg resource.Managed) error {
 
 	fmt.Printf("Deleting: %+v", cr)
 
-	cr.Status.SetConditions(runtimev1alpha1.Deleting())
+	cr.Status.SetConditions(xpv1.Deleting())
 
 	return nil
 }
+
+func ReconcileLogic(conditionAfter []v1alpha1.ResourceConditionAfter, t time.Duration) []int {
```
Comment (jbw976) [inline internal/controller/nopresource/nopresource.go]: because this function is UpperCase, it will be an exported function.  There's no reason to export this function since it's only called within this package, so you can call it `reconcileLogic` instead to keep it from being exported.

File: internal/controller/nopresource/nopresource.go [hunk 8]
```diff
@@ -108,57 +108,29 @@ func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
...
-		fmt.Printf("Calling update on index %d\n", l)
-
-		x := runtimev1alpha1.Condition{
-			Type:               runtimev1alpha1.ConditionType(cr.Spec.ForProvider.ConditionAfter[l].ConditionType),
+		x := xpv1.Condition{
+			Type:               xpv1.ConditionType(cr.Spec.ForProvider.ConditionAfter[l].ConditionType),
 			Status:             v1.ConditionStatus(cr.Spec.ForProvider.ConditionAfter[l].ConditionStatus),
 			LastTransitionTime: metav1.Now(),
-			Reason:             runtimev1alpha1.ReasonAvailable,
+			Reason:             xpv1.ReasonAvailable,
```
Comment (jbw976) [inline internal/controller/nopresource/nopresource.go:117]: this Reason is not always going to be accurate depending on what the condition actually is - Ready (true/false) and Synced (true/false) all have different reasons associated with them: https://github.com/crossplane/crossplane-runtime/blob/master/apis/common/v1/condition.go#L43-L54

File: internal/controller/nopresource/nopresource.go [hunk 9]
```diff
@@ -217,7 +189,29 @@ func (c *external) Delete(ctx context.Context, mg resource.Managed) error {
...
-	cr.Status.SetConditions(runtimev1alpha1.Deleting())
+	cr.Status.SetConditions(xpv1.Deleting())
 
 	return nil
 }
+
+func ReconcileLogic(conditionAfter []v1alpha1.ResourceConditionAfter, t time.Duration) []int {
+	latestTime := make(map[string]time.Duration)
+	latestIdx := make(map[string]int)
+	for i := 0; i < len(conditionAfter); i++ {
```
Comment (jbw976) [inline internal/controller/nopresource/nopresource.go:200]: this function looks like it's logically correct, but it could use a bit of clean up in the form of comments and variable names.  e.g., the function header should be commented to be clear about what the arguments mean, how they should be used, and what the return value means (a slice of ints that represents what exactly).  Some of the checks within the body of the function could use a comment saying in plain english what it is checking for.

The variable names could be cleaned up (more descriptive names) as i was getting a bit confused between `t`, `mx` and `elapsed`.

File: internal/controller/nopresource/nopresource_test.go [hunk 10]
```diff
@@ -72,3 +75,79 @@ func TestObserve(t *testing.T) {
 		})
 	}
 }
+
+func TestReconcileLogic(t *testing.T) {
+
+	c := []v1alpha1.ResourceConditionAfter{
```
Comment (jbw976) [inline internal/controller/nopresource/nopresource_test.go:82]: i like that this test slice has multiple entries at the same time and that it's not in order either.  good test cases!

File: internal/controller/nopresource/nopresource_test.go [hunk 11]
```diff
@@ -72,3 +75,79 @@ func TestObserve(t *testing.T) {
...
+			elapsedtime:       time.Duration(1*time.Second + 999*time.Millisecond),
+			want:              []int{},
+		},
+		"SingleTypeReconcile": {
+			reason:            "Slice with a single element should be returned when a single condition type has been specified.",
+			resourcecondition: c,
+			elapsedtime:       time.Duration(2 * time.Second),
+			want:              []int{5},
+		},
+		"NormalReconcileBehaviour": {
```
Comment (jbw976) [inline internal/controller/nopresource/nopresource_test.go:115]: probably another test case after 5 sec to test that the condition of Ready: false is still activated

File: internal/controller/nopresource/nopresource_test.go [hunk 12]
```diff
@@ -72,3 +75,79 @@ func TestObserve(t *testing.T) {
...
+			if !sameIntSlice(tc.want, got) {
+				t.Errorf("\n%s\nReconcileLogic(...): want []int %v, got []int %v\n", tc.reason, tc.want, got)
+			}
+		})
+	}
+
+}
+
+// Matches the content of the slices. Order does not matter.
+func sameIntSlice(x, y []int) bool {
```
Comment (jbw976) [inline internal/controller/nopresource/nopresource_test.go]: i wouldn't think you'd have to implement this comparison yourself - look into this `cmpopts.SortSlices` test func we use in other repos in the crossplane org - that looks relevant for what you want to do with int slice comparisons:

https://github.com/crossplane/crossplane-runtime/blob/master/apis/common/v1/condition.go#L43-L54
